### ISSUE TITLE
第一次双周报告

### ISSUE THREAD

**Issue by Lfan-ke** [Lfan-ke/ES-MoE#5]
## 工作概述

三个调用把 ES-MoE 块接进任意一份官方 YOLO 配置：`inject_esmoe` 注册该块，`graft` 插入并修正层引用，`attach_aux_loss` 把路由的负载均衡损失接进优化器；`equip` 一次跑完。

    pip install esmoe

    import esmoe
    model = esmoe.equip("yolo11n.yaml", weight=0.01)
    model.train(data="VisDrone.yaml", epochs=20)

由三项单测可证辅助损失确已进入反向传播：其一，开启与关闭该项时，总损失之差恰为其数值乘以 `batch_size`；其二，路由器参数确有梯度；其三，`results.csv` 中 `esmoe_aux` 自成一列且数值非零。配置中写有 `aux_loss` 键并不构成证据。

## 验收相关

| 要求 | 现状 |
|:--:|:--:|
| 插件包 | [`pip install esmoe`](https://pypi.org/project/esmoe/)，0.1.0 → 0.1.3 |
| 兼容矩阵 | YOLOv8n / YOLO11n / YOLO12n / YOLO26n 同规格各跑 1 epoch，`train/esmoe_aux` 均非零 |
| 多 seed 消融 | 四组配对实验与五组配置选型，[results/](https://github.com/Lfan-ke/ES-MoE/tree/main/results) 共 40 条 JSON 记录 |
| README / Colab | [文档站](https://lfan-ke.github.io/ES-MoE/)中英双语，快速上手可在 Colab 直接运行 |
| 发布说明 | 每个版本一份，随文档站发布 |

## 首轮数字

全量 VisDrone，开启与关闭各三个 seed，同数据、同预算、同增广、同评测口径；比的是同 seed 的配对差值，而非两组均值。

| 组 | baseline mAP50 | esmoe mAP50 | 逐 seed 差 | 均值 | 胜 |
|:--:|:--:|:--:|:--:|:--:|:--:|
| YOLOv8n · 20 ep | 0.1496 | 0.1517 | +0.0001 / +0.0032 / +0.0029 | +0.0021 | 3/3 |
| YOLOv8n · 50 ep | 0.2131 | 0.2143 | −0.0003 / +0.0009 / +0.0032 | +0.0013 | 2/3 |
| YOLOv8n · 100 ep | 0.3004 | 0.3050 | +0.0104 / −0.0011 / +0.0044 | +0.0046 | 2/3 |
| YOLO11n · 20 ep | 0.1419 | 0.1410 | +0.0045 / −0.0033 / −0.0040 | −0.0009 | 1/3 |

三个预算的均值都为正，但 seed 之间的离散度随预算增大，单次运行不足以支撑结论；换到 YOLO11n 主干上效应不成立。代价是参数增加 10.4%，每轮墙钟增加 2%~13%。

兼容矩阵证的是机制而非精度：四代主干都能建、能训，辅助损失也都进了总损失；精度结论目前只覆盖 YOLOv8n。

此前只有 20 与 50 两个预算的数据，据此写过"增益随训练变长而衰减"；100 epoch 的结果不支持这个说法，文档已改写。

## 遇到的问题

- **训练器会从配置重建模型，EMA 副本又在所有 callback 之前生成**：挂在模型实例上的辅助损失权重会丢，只能放进程作用域。
- **`loss_items` 跨 Ultralytics 版本换过形态**：8.4.13x 之前是 `tensor`，之后是具名 `dict`，两种都要处理；`loss_names` 也只能在非空时追加，否则日志表头错位。
- **追踪导出会固化路由**：`torch.jit.trace` 只记录当前这一批走过的分支，所得的图对其余输入一律按同一条路由走，结果错误，且不报错；守卫须写成 `torch.jit.is_tracing() or torch.onnx.is_in_onnx_export()`。

## 下周计划

做关键消融与中期演示，补单测与文档。其中要弄清的是效应为何不迁移到 YOLO11n：是插入点不合适，还是块本身的问题。收敛行为、COCO 与其他模型规模均尚未测试，三个 seed 也不足以谈显著性，完整清单见[已知局限](https://lfan-ke.github.io/ES-MoE/limitations/)。

---

排查上述追踪固化问题时，发现上游 `OptimizedMOE` / `OptimizedMOEImproved` 只挡了 ONNX 导出、没挡 `torch.jit.trace`，已带回归测试提 PR [Tencent/YOLO-Master#241](https://github.com/Tencent/YOLO-Master/pull/241)，已合并。
